pkg/networkservicemesh/common/heal/client.go

```diff
@@ -0,0 +1,174 @@
+package heal
+
+import (
+	"context"
+	"runtime"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/connection"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/networkservice"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservicemesh/core/trace"
+	"github.com/networkservicemesh/sdk/pkg/tools/extend"
+	"github.com/networkservicemesh/sdk/pkg/tools/serialize"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservicemesh/core/next"
+)
+
+type healClient struct {
+	requestors        map[string]func()
+	closers           map[string]func()
+	reported          map[string]*connection.Connection
+	onHeal            *networkservice.NetworkServiceClient
+	client            connection.MonitorConnectionClient
+	eventReceiver     connection.MonitorConnection_MonitorConnectionsClient
+	updateExecutor    serialize.Executor
+	recvEventExecutor serialize.Executor
+	cancelFunc        context.CancelFunc
+}
+
+// NewClient - creates a new networkservice.NetworkServiceClient chain element that implements the healing algorithm
+//             - client - connection.MonitorConnectionClient that can be used to call MonitorConnection against the endpoint
+//             - onHeal - *networkservice.NetworkServiceClient.  Since networkservice.NetworkServiceClient is an interface
+//                        (and thus a pointer) *networkservice.NetworkServiceClient is a double pointer.  Meaning it
+//                        points to a place that points to a place that implements networkservice.NetworkServiceClient
+//                        This is done because when we use heal.NewClient as part of a chain, we may not *have*
+//                        a pointer to this
+//                        client used 'onHeal'.  If we detect we need to heal, onHeal.Request is used to heal.
+//                        If onHeal is nil, then we simply set onHeal to this client chain element
+//                        If we are part of a larger chain or a server, we should pass the resulting chain into
+//                        this constructor before we actually have a pointer to it.
+//                        If onHeal nil, onHeal will be pointed to the returned networkservice.NetworkServiceClient
+func NewClient(client connection.MonitorConnectionClient, onHeal *networkservice.NetworkServiceClient) networkservice.NetworkServiceClient {
+	rv := &healClient{
+		onHeal:            onHeal,
+		requestors:        make(map[string]func()),
+		closers:           make(map[string]func()),
+		reported:          make(map[string]*connection.Connection),
+		client:            client,
+		updateExecutor:    serialize.NewExecutor(),
+		eventReceiver:     nil,                     // This is intentionally nil
+		recvEventExecutor: serialize.NewExecutor(), // This is intentionally nil
```

Comment:
should this be a serialize.NewExecutor() or nil? I suspect the comment is incorrect due to usage further down.

Reply:
Its intentionally nil.  populating eventReceiver involves making at *tcp* call... we want to minimize startup latency, so the tcp call happens in a updateExcutor.AsyncExec below...